### PR TITLE
Add support for mounting multiple base folders

### PR DESCRIPTION
**Pull request by oldmud0**
This feature introduces the ability for players to mount multiple base folders in addition to the main AO base folder. When an asset is not found in the main base folder, the list of mounted base folders is searched in order to find the asset.

This can be used to split up assets belonging to various servers and configure an order of precedence when two character folders conflict. It can also be used to store a small subset of characters locally, and then have the rest of the characters be located on another drive or even a network location (as long as it is mounted as an OS drive).

This feature does not introduce any breaking changes. Assets with ambiguous extensions are loaded in the way you would expect them to (try all extensions before moving on to the next mount path). Only setting files (e.g. callwords.ini, serverlist.txt) will continue to be limited to the main base folder.

### CODE REVIEW

**Review by in1tiate — DISMISSED** [review]
comes with some welcome improvements to path resolution, i am a fan of this

File: include/aoapplication.h [hunk 1]
```diff
@@ -37,6 +37,25 @@ class NetworkManager;
...
+  inline VPath operator+(const VPath &str) const {
+    return VPath(this->toQString() + str.toQString());
+  }
+};
+
+inline uint qHash(const VPath &key, uint seed)
+{
+  return qHash(key.toQString(), seed);
+}
+
```
Comment (Crystalwarrior) [inline include/aoapplication.h:58]: what is this black magic
holy  crap, ao2 using actual c++

File: include/aoapplication.h [hunk 2]
```diff
@@ -130,24 +149,26 @@ class AOApplication : public QApplication {
...
+  QString get_asset_path(QVector<VPath> pathlist);
+  QString get_image_path(QVector<VPath> pathlist, bool static_image=false);
+  QString get_sfx_path(QVector<VPath> pathlist);
   QString get_config_value(QString p_identifier, QString p_config, QString p_theme="", QString p_subtheme="", QString p_default_theme="", QString p_misc="");
   QString get_asset(QString p_element, QString p_theme="", QString p_subtheme="", QString p_default_theme="", QString p_misc="", QString p_character="", QString p_placeholder="");
-  QString get_image(QString p_element, QString p_theme="", QString p_subtheme="", QString p_default_theme="", QString p_misc="", QString p_character="", QString p_placeholder="");
+  QString get_image(QString p_element, QString p_theme="", QString p_subtheme="", QString p_default_theme="", QString p_misc="", QString p_character="", QString p_placeholder="", bool static_image=false);
   QString get_sfx(QString p_sfx, QString p_misc="", QString p_character="");
   QString get_case_sensitive_path(QString p_file);
+  QString get_real_path(const VPath &vpath);
```
Comment (Crystalwarrior) [inline include/aoapplication.h:169]: get_real()

File: src/aoapplication.cpp [hunk 3]
```diff
@@ -18,7 +18,8 @@ AOApplication::AOApplication(int &argc, char **argv) : QApplication(argc, argv)
   discord = new AttorneyOnline::Discord();
   QObject::connect(net_manager, SIGNAL(ms_connect_finished(bool, bool)),
                    SLOT(ms_connect_finished(bool, bool)));
-  qApp->setStyleSheet("QFrame {background-color:transparent;} QAbstractItemView {background-color: transparent; color: black;}; QLineEdit {background-color:transparent;}");
+  // qApp->setStyleSheet("QFrame {background-color:transparent;} QAbstractItemView {background-color: transparent; color: black;}; QLineEdit {background-color:transparent;}");
```
Comment (Crystalwarrior) [inline src/aoapplication.cpp:21]: Why was this commented out? This was added to make sure that outdated themes with .css don't break old behavior.

**Review by Crystalwarrior — CHANGES_REQUESTED** [review]
Lots of sweeping changes across path resolution in this PR, you should make sure to test that all of the pre-existing behavior is not broken. Perhaps a pre-release can be put out so people can report bugs swiftly

Also, dunno why you commented out default stylesheet settings for AO App object
Cache is cool and good though, there's some giga brain stuff in here I don't fully understand lol

**Comment by in1tiate** [comment]
needs more testing eh? one hell of a change of pace from pre-2.9.1

File: src/aoapplication.cpp (hunk 3 above)
Comment (Crystalwarrior) [reply src/aoapplication.cpp:21]: Set it on the courtroom itself maybe rather than the app?

Comment (Crystalwarrior) [reply src/aoapplication.cpp:21]: Set Default style sheet on the court and we Gucci

**Review by Crystalwarrior — CHANGES_REQUESTED** [review]
Set Default style sheet on the court and we Gucci

**Comment by oldmud0** [comment]
@Crystalwarrior gucci?

**Comment by Crystalwarrior** [comment]
> @Crystalwarrior we gucci?

I said court not lobby. Just do both

**Review by Crystalwarrior — APPROVED** [review]
Ye

**Comment by oldmud0** [comment]
yey

**Pull request merged**